File: connector/connector-api/src/main/java/com/github/thetric/iliasdownloader/service/IliasService.java
package com.github.thetric.iliasdownloader.service;

import com.github.thetric.iliasdownloader.service.model.Course;
import com.github.thetric.iliasdownloader.service.model.CourseFile;
import com.github.thetric.iliasdownloader.service.model.CourseItem;
import com.github.thetric.iliasdownloader.service.model.LoginCredentials;
import io.reactivex.Observable;
import io.reactivex.Single;

import java.util.Collection;

/**
 * Provides access to the Ilias.
 * This interface provides some methods for basic session management (login, logout) and it can list the courses
 * (with/without their contents).
 *
 * @author broj
 * @since 21.05.2016
 */
public interface IliasService {

    // Session management

    /**
     * Logs the user in.
     *
     * @param loginCredentials
     *         user credentials
     * @see #logout()
     */
    void login(LoginCredentials loginCredentials);

    /**
     * Logs the current user out.
     *
     * @see #login(LoginCredentials)
     */
    void logout();

    // course sync

    /**
     * Finds all courses without any course content.
     *
     * @return all courses of the current user
     * @see #searchCoursesWithContent(Collection)
     */
    Observable<Course> getJoinedCourses();

    /**
     * Searches the selected courses with their child nodes <b>without downloading them</b>.
     *
     * @param selectedCourses
     *         {@link Course}s to search for. The list itself and its items <i>must not</i> be modified under any
     *         circumstances
     * @return new list with {@link Course}s and their child nodes
     * @see CourseItem
     */
    Observable<Course> searchCoursesWithContent(Collection<Course> selectedCourses);

    /**
     * Downloads the content of the {@link CourseFile} from the Ilias.
     *
     * @param file
     *         {@link CourseFile} to download
     * @return {@link Single} which publishes either the data as a byte array or an exception
     * @see java.io.IOException
     * @see com.github.thetric.iliasdownloader.service.exception.IliasException
     */
    Single<byte[]> getContent(CourseFile file);
}

File: src/main/kotlin/com/github/thetric/iliasdownloader/connector/api/IliasService.kt
package com.github.thetric.iliasdownloader.connector.api

import com.github.thetric.iliasdownloader.connector.api.model.Course
import com.github.thetric.iliasdownloader.connector.api.model.CourseFile
import com.github.thetric.iliasdownloader.connector.api.model.LoginCredentials
import java.io.IOException
import java.io.InputStream

/**
 * Provides access to the Ilias.
 * This interface provides some methods for basic session management (login,
 * logout) and it can list the courses (with/without their contents).
 */
interface IliasService {
    // Session management

    /**
     * Logs the user in.
     *
     * @param loginCredentials user credentials
     * @see [logout]
     */
    fun login(loginCredentials: LoginCredentials)

    /**
     * Logs the current user out.
     *
     * @see [login]
     */
    fun logout()

    // course sync

    /**
     * Finds all courses without any course content.
     *
     * @return all [Course]s of the current user
     * @see [visit]
     */
    fun getJoinedCourses(): Collection<Course>

    fun <C> visit(
        courseItem: Course,
        itemVisitor: ContextAwareIliasItemVisitor<C>,
        initialContext: C
    )

    /**
     * Downloads the content of the {@link CourseFile} from the Ilias.
     *
     * @param file [CourseFile] to download
     * @return [InputStream] content of the file
     * @see com.github.thetric.iliasdownloader.connector.api.exception.IliasException
     */
    @Throws(IOException::class)
    fun getContentAsStream(file: CourseFile): InputStream
}
